kernel/src/scan/mod.rs

```diff
@@ -384,6 +384,78 @@ fn parse_partition_value(raw: Option<&String>, data_type: &DataType) -> DeltaRes
     }
 }
 
+/// Transform a logical field into the physical form. Currently just renames things for 'name'
+/// column mapping.
+fn make_field_physical(
+    logical_field: &StructField,
+    column_mapping_mode: ColumnMappingMode,
+) -> DeltaResult<StructField> {
+    match column_mapping_mode {
+        ColumnMappingMode::None => Ok(logical_field.clone()),
+        ColumnMappingMode::Name => {
+            let physical_name = logical_field.physical_name(column_mapping_mode)?;
```

Comment:
Column mapping is pretty fundamental... I wonder if we should define these as methods on the various types?

<details>

```rust
impl StructField {
    pub fn to_physical(self, mode: ColumnMappingMode) -> DeltaResult<Self> {
        Ok(Self {
            name: self.physical_name(mode)?.into(),
            data_type: self.data_type.to_physical(mode)?,
            ..self,
        })
    }
}

impl DataType {
    pub fn to_physical(self, mode: ColumnMappingMode) -> DeltaResult<Self> {
        let result = match self {
            Struct(s) => DataType::try_struct_type(
                s.fields.into_values().map(|field| field.to_physical(mode)),
            )?,
            Array(a) => DataType::array_type(
                a.element_type.to_physical(mode)?,
                a.contains_null,
            ),
            Map(m) => DataType::map_type(
                m.key_type.to_physical(mode)?,
                m.value_type.to_physical(mode)?,
                m.value_contains_null,
            ),
            Primitive(_) => self,
        };
        Ok(result)
    }
}
```

Note that the above consumes `self` so we would need to `clone` the input first. Seems like a welcome simplification that doesn't change big-O bound and so satisfies the "don't optimize, don't pessimize" mantra?

</details>

Meanwhile: The existing `StructField::with_field` takes `&self` instead of `self` like `with_XXX` methods normally do. It only has one call site so far (ignoring this PR), so we should probably fix that ASAP.

Also, the existing `StructType::new` takes `Vec<StructField>` when it would be much nicer to take `impl IntoIterator<Item = StructField>`. I'll probably throw up a PR to fix that one soon.

Also, I noticed that the `test_field_metadata` unit test (schema.rs) for physical names is going directly to the metadata instead of exercising the `physical_name` method we actually use in practice??

Reply:
PR to simplify `StructType` is up: https://github.com/delta-incubator/delta-kernel-rs/pull/385

Reply:
Given your newer PR https://github.com/delta-incubator/delta-kernel-rs/pull/395 I'll hold off on responding here

Reply:
Resolving as this is handled on the type